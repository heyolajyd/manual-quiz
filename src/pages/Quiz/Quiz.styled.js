import styled from 'styled-components';
import { breakpoints } from 'styles/global';

export const QuizContainer = styled.section`
  border: 2px solid ${(props) => props.theme.colors.tealLight};
  display: flex;
  flex-direction: column;
  min-height: 650px;
  margin: 50px auto;
  width: 650px;

  @media ${breakpoints.sm} {
    width: 90%;
    margin: 20px auto;
  }
`;

export const QuizContent = styled.section`
  display: flex;
  flex-grow: 1;
  flex-direction: column;
  padding: 40px 60px;
  text-align: left;

  @media ${breakpoints.sm} {
    padding: 15px;
  }
`;
