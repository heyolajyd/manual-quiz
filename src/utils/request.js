/**
 * Util function to get questions
 */
export const getQuestion = async () => {
  try {
    const res = await fetch('./questions.json');
    return await res.json();
  } catch (err) {
    console.log('Failed to parse: ', JSON.stringify(err));
    throw new Error(err);
  }
};
