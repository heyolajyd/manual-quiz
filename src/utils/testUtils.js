import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render } from '@testing-library/react';
import { theme } from 'styles/global';

export const withThemeProvider = (component) =>
  render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);
