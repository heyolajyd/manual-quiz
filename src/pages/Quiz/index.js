import React, { useEffect, useState } from 'react';
import { getQuestion } from 'utils/request';
import ProgressBar from 'components/ProgressBar';
import Result from 'components/Result';
import Header from 'components/Header';
import Question from 'components/Question';
import { QuizContainer, QuizContent } from './Quiz.styled';

const Quiz = () => {
  const [questions, setQuestions] = useState([]);
  const [currentQuestion, setCurrentQuestion] = useState({});
  const [currentNumber, setCurrentNumber] = useState(0);
  const [completedQuiz, setCompletedQuiz] = useState(false);
  const [selectedAnswers, setSelectedAnswers] = useState([]);

  useEffect(() => {
    fetchQuestions();
  }, []);

  useEffect(() => {
    if (questions.length) {
      const question = questions[currentNumber];
      setCurrentQuestion(question);
    }
  }, [questions, currentNumber]);

  const initializeState = () => {
    setQuestions([]);
    setCurrentQuestion({});
    setCurrentNumber(0);
    setCompletedQuiz(false);
    setSelectedAnswers([]);
  };

  const fetchQuestions = async () => {
    const { questions: quizQuestions } = await getQuestion();

    setQuestions(quizQuestions);
  };

  const handleNextQuestion = (answer) => {
    setSelectedAnswers([...selectedAnswers, answer]);

    const nextNumber = currentNumber + 1;
    return nextNumber < questionCount
      ? setCurrentNumber(nextNumber)
      : setCompletedQuiz(true);
  };

  const handleStartQuiz = () => {
    initializeState();
    fetchQuestions();
  };

  const questionCount = questions.length;

  return (
    <QuizContainer>
      <ProgressBar
        questionCount={questionCount}
        questionNumber={currentNumber + 1}
      />
      <QuizContent>
        {completedQuiz ? (
          <Result
            isSupportAvailable={!selectedAnswers.includes('true')}
            startQuiz={handleStartQuiz}
          />
        ) : (
          <>
            <Header
              questionNumber={currentNumber}
              questionCount={questionCount}
            />
            <Question
              currentQuestion={currentQuestion}
              nextQuestion={handleNextQuestion}
            />
          </>
        )}
      </QuizContent>
    </QuizContainer>
  );
};

export default Quiz;
