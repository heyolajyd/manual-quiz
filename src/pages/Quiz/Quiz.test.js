import React from 'react';
import { withThemeProvider } from 'utils/testUtils';
import Quiz from './index';

describe('Quiz page', function () {
  it('should match snapshot', async () => {
    const { container } = withThemeProvider(<Quiz />);

    expect(container).toMatchSnapshot();
  });

  it('should display question', async () => {
    const { getByText } = withThemeProvider(<Quiz />);

    expect(getByText(/Question/)).toBeInTheDocument();
  });
});
