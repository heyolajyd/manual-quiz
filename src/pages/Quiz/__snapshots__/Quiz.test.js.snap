// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz page should match snapshot 1`] = `
<div>
  <section
    class="sc-fzplWN iwfOcV"
  >
    <div
      class="sc-AxjAm hUTlVe"
    >
      <div
        class="sc-AxirZ fCefnb"
        width="0"
      />
    </div>
    <section
      class="sc-fznyAO bCqqLA"
    >
      <div
        class="sc-AxgMl fFPQpl"
      >
        <h1
          class="sc-AxheI UZKLG"
        >
          Question 
          1
           of 
          0
        </h1>
      </div>
      <section
        class="sc-Axmtr KZRLv"
      >
        <div
          class="sc-AxmLO hFmTzr"
        />
        <ul
          class="sc-fzozJi jANZmt"
        />
      </section>
    </section>
  </section>
</div>
`;
